### PR TITLE
Move file storage extension to core

### PR DESCRIPTION
**Pull request by pmm-sumo**
**Description:**

This is a copy of [file storage extension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage/filestorage) moved into the core, which might be required for file-storage extension backed WAL implementation

### CODE REVIEW

File: extension/storage/Makefile [hunk 1]
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Comment (bogdandrutu) [inline extension/storage/Makefile:1]: Why this?

File: extension/storage/filestorage/Makefile [hunk 2]
```diff
@@ -0,0 +1 @@
+include ../../Makefile.Common
```
Comment (bogdandrutu) [inline extension/storage/filestorage/Makefile:1]: Why this change?

File: extension/storage/filestorage/extension.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func (lfs *localFileStorage) GetClient(ctx context.Context, kind component.Kind, ent config.ComponentID) (storage.Client, error) {
+	rawName := fmt.Sprintf("%s_%s_%s", kindString(kind), ent.Type(), ent.Name())
+	// TODO sanitize rawName
+	absoluteName := filepath.Join(lfs.directory, rawName)
+
+	client, err := newClient(absoluteName, lfs.timeout)
+	if err != nil {
+		return nil, fmt.Errorf("create client: %v", err)
+	}
+	lfs.clients = append(lfs.clients, client)
```
Comment (bogdandrutu) [inline extension/storage/filestorage/extension.go:80]: Is it guaranteed that GetClient is not called by multiple goroutines? You need to protect the slice.

File: extension/storage/filestorage/extension.go [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+// Start does nothing
+func (lfs *localFileStorage) Start(context.Context, component.Host) error {
+	return nil
+}
+
+// Shutdown will close any open databases
+func (lfs *localFileStorage) Shutdown(context.Context) error {
+	for _, client := range lfs.clients {
+		client.close()
+	}
```
Comment (bogdandrutu) [inline extension/storage/filestorage/extension.go:64]: Sounds like a chaos here, there is no clear owner of the components. Which component is responsible to close the "client" instance? Not well documented and based on the code seems a bit strange to say the least.

**Review by bogdandrutu — COMMENTED** [review]
That is the motivation of this change?

**Comment by pmm-sumo** [comment]
Thanks @bogdandrutu , ;et me give a bit of context. I am researching how to implement WAL in queued_retry (and then other components, batch processor, group by trace processor, etc.) and one of the ideas is to leverage [storage](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/storage) extension which is already part of `opentelemetry-collector-contrib`. The idea behind this extension was to make it generally available to other components and it's largely happening here. The PR for WAL implementation in queued_retry using storage extension is here: https://github.com/open-telemetry/opentelemetry-collector/pull/3274

In this PR, I just wanted to show what would be the scope of bringing storage extension to the core, without changing the structure of it

File: extension/storage/Makefile (hunk 1 above)
Comment (djaglowski) [reply extension/storage/Makefile:1]: Artifact from being a contrib component. Should be removed in core implementation.

File: extension/storage/filestorage/Makefile (hunk 2 above)
Comment (djaglowski) [reply extension/storage/filestorage/Makefile:1]: Should be removed.

File: extension/storage/filestorage/extension.go (hunk 4 above)
Comment (djaglowski) [reply extension/storage/filestorage/extension.go:64]: Yeah, fair point. 

When I implemented this, I had in mind the concept of cleaning up stale storage files [#3149 (contrib)](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/3149), which I think would properly be the responsibility of this extension and might look similar to this implementation of tracking/closing clients. 

In any case, you are right - the clients can and should be handled by the components that requested them, and this expectation should be clearly documented.

File: extension/storage/filestorage/extension.go (hunk 3 above)
Comment (djaglowski) [reply extension/storage/filestorage/extension.go:80]: This slice is only used for client closure, which will become the responsibility of each component that requests a client. So this will go away entirely. Point taken though.

**Review by djaglowski — COMMENTED** [review]
@pmm-sumo I will make the noted change on a contrib PR. Please let me know if you'd like to me to PR to your branch here as well.

**Comment by pmm-sumo** [comment]
> @pmm-sumo I will make the noted change on a contrib PR. Please let me know if you'd like to me to PR to your branch here as well.

Thank you Dan! I can take care of that later and just sync the changes. Also, if you prefer the prepare the PR with filestorage changes to _core_ (which I think you deserve, as the author) :) then it works perfectly fine for me too. I can adjust my PR with WAL implementation later accordingly to either of options.
